### PR TITLE
add moving average / percentiles to charge level and support longer periods

### PR DESCRIPTION
**Pull request by swiffer**
Successor of #4187 

added several parameters to allow flexible sizing of bucket width & option to disable the percentiles

**before:**
charge-level

**after:**
grafik

### CODE REVIEW

**Comment by swiffer** [comment]
@JakobLichterfeld - ok, should be ready for testing!

File: grafana/dashboards/charge-level.json [hunk 1]
```diff
@@ -225,6 +243,35 @@
...
...ted_gapfilled_locf\r\n    window w as (rows between (86400 / $bucket_width) * ($days_moving_average_percentiles / 2) preceding and (86400 / $bucket_width) * ($days_moving_average_percentiles / 2) following)\r\n)\r\n\r\nselect\r\n    series_id,\r\n    case when battery_level is null then null else (select percentile_cont(0.075) within group (order by s) from unnest(arr) trick(s)) end as \"$days_moving_average_percentiles Day Moving 7,5% Percentile (${bucket_width:text} buckets)\",\r\n    case when battery_level is null then null else (battery_level_avg) end as \"$days_moving_average_percentiles Day Moving Average (${bucket_width:text} buckets)\",\r\n    case when battery_level is null then null else (select percentile_cont(0.5) within group (order by s) from unnest(arr) trick(s)) end as \"$days_moving_average_percentiles Day Moving Median (${bucket_width:text} buckets)\",\r\n    case when battery_level is null then null else (select percentile_cont(0.925) within group (order by s) from unnest(arr) trick(s)) end as \"$days_moving_average_percentiles Day Moving 92,5% Percentile (${bucket_width:text} buckets)\"\r\nfrom\r\n    positions_bucketed_gapfilled_locf_percentile_intermediate;",
```
Comment (JakobLichterfeld) [inline grafana/dashboards/charge-level.json:256]: `calucate` -> `calculate`
and 
`calcuations` -> `calculations`

**Review by JakobLichterfeld — CHANGES_REQUESTED** [review]
Thanks so much!

File: grafana/dashboards/charge-level.json (hunk 1 above)
Comment (swiffer) [reply grafana/dashboards/charge-level.json]: fixed

**Review by JakobLichterfeld — APPROVED** [review]

**Comment by swiffer** [comment]
@JakobLichterfeld - will add a small follow up PR - need to increase generate series to correctly calculate the average for the first 15 days...

**Comment by JakobLichterfeld** [comment]
> will add a small follow up PR - need to increase generate series to correctly calculate the average for the first 15 days...

no worries, love the eagle eyes.

**Comment by swiffer** [comment]
I don't know. First interesting to see how charge level / battery usage changes over time and how much capacity is used usually... would be great to compare curves with other users and ensure percentiles always stay within the green lines. No deep knowledge about recommended charge levels here, sorry.

**Comment by jim0266** [comment]
I don't understand the changes made to the charge level. It no longer displays the actual charge level of my car. How is this better? What am I suppose to glean from this data? Is there any documentation explaining the changes to charge level? Thanks.

**Comment by swiffer** [comment]
Hi @jim0266 - you are not seeing the yellow line anymore?

grafik

**Comment by jim0266** [comment]
Screenshot 2024-11-19 at 10 23 32 AM

Hopefully this screenshot helps. From 9/4 to 9/14 we took at 3,300 mile trip. We charged as high as 90% on 9/5 and went as low as 16% later in the trip. None of that is captured in my data or I don't know how to configure it now to show that.

How can I see this accurate data? I'm not sure what use the current data is to me. What is it trying to tell me?

I'm also showing a DB error is this is part of the problem.

Thanks for any help or insights.

**Comment by swiffer** [comment]
you need to upgrade your postgres instance - see here for instructions - and ensure you have a backup!

https://docs.teslamate.org/docs/maintenance/upgrading_postgres

**Comment by jim0266** [comment]
All was good until this last step:

root@teslaserver:~# docker compose start teslamate
WARN[0000] The "apr1" variable is not set. Defaulting to a blank string. 
WARN[0000] The "vp9ymblq" variable is not set. Defaulting to a blank string. 
WARN[0000] The "o6stCrEwF1q3cnC7pD1y2" variable is not set. Defaulting to a blank string. 
WARN[0000] The "apr1" variable is not set. Defaulting to a blank string. 
WARN[0000] The "vp9ymblq" variable is not set. Defaulting to a blank string. 
WARN[0000] The "o6stCrEwF1q3cnC7pD1y2" variable is not set. Defaulting to a blank string. 
WARN[0000] /root/docker-compose.yml: `version` is obsolete 

For the version I used 17.1, which is the latest stable version of Postgres. I am hosted at Vultr.

 database:
    image: postgres:17.1

Thanks!

**Comment by swiffer** [comment]
can you compare the docker compose file with the one mentioned in the docs or post it here (redacted (passwords)).

**Comment by JakobLichterfeld** [comment]
Please open a separate Ticket, explain what you have done, post your docker-compose.yml (without secrets)

**Pull request merged**